resources/src/effect/mod.rs

```diff
@@ -0,0 +1,564 @@
+use amount::Amount;
+use asset::{AssetIdentifier, Flag};
+use serde::{de, Deserialize, Deserializer};
+
+pub mod account;
+pub mod signer;
+pub mod trustline;
+mod trade;
+
+pub use self::trade::Trade;
+
+#[cfg(test)]
+mod test;
+
+/// A successful operation will yield zero or more effects. These effects represent specific
+/// changes that occur in the ledger, but are not necessarily directly reflected in the ledger or
+/// history, as transactions and operations are.
+#[derive(Debug)]
+pub struct Effect {
+    id: String,
+    paging_token: String,
+    kind: Kind,
+}
+
+/// Each effect type is representing by a kind and captures data specific to that
+/// type within it's newtype.
+#[derive(Debug, Deserialize)]
+pub enum EffectKind {
+    /// A collection of effects that represent updates to an account
+    Account(account::Kind),
+    /// A collection of effects that represent updates to an account signer
+    Signer(signer::Kind),
+    /// A collection of effects that represent updates to a trustline
+    Trustline(trustline::Kind),
+    /// An effect representing a trade being executed
+    Trade(Trade),
+    // The stellar api docs list other operations for offers, but as of this writing those
+    // endpoints do not yet exist in horizon https://github.com/stellar/go/issues/166
+}
+// Use inside file to be brief
+use self::EffectKind as Kind;
+
+impl Effect {
+    /// the unique identifier of an effect
+    pub fn id(&self) -> &String {
+        &self.id
+    }
+
+    /// A paging token suitable for use as a cursor parameter.
+    pub fn paging_token(&self) -> &String {
+        &self.paging_token
+    }
+
+    /// Specifies the type of effect, See “Types” section below for reference.
+    pub fn type_i(&self) -> u32 {
+        match &self.kind {
+            &Kind::Account(ref account_kind) => match account_kind {
```

Comment:
You could capture the inner kind by ref and then deref it for the match. This removes all the little ampersands:

```rust

/// Specifies the type of effect, See “Types” section below for reference.
pub fn type_i(&self) -> u32 {
    match self.kind {
        Kind::Account(ref account_kind) => match *account_kind {
            account::Kind::Created(_) => 0,
            account::Kind::Removed(_) => 1,
            account::Kind::Credited(_) => 2,
            account::Kind::Debited(_) => 3,
            account::Kind::ThresholdsUpdated(_) => 4,
            account::Kind::HomeDomainUpdated(_) => 5,
            account::Kind::FlagsUpdated(_) => 6,
        },
        Kind::Signer(ref signer_kind) => match *signer_kind {
            signer::Kind::Created(_) => 10,
            signer::Kind::Removed(_) => 11,
            signer::Kind::Updated(_) => 12,
        },
        Kind::Trustline(ref trustline_kind) => match *trustline_kind {
            trustline::Kind::Created(_) => 20,
            trustline::Kind::Removed(_) => 21,
            trustline::Kind::Updated(_) => 22,
            trustline::Kind::Authorized(_) => 23,
            trustline::Kind::Deauthorized(_) => 24,
        },
        Kind::Trade(_) => 33,
    }
}
```